cli_tools/common/utils/daisyutils/workflow_hook_fallback_pd.go

```diff
@@ -0,0 +1,62 @@
+//  Copyright 2021 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package daisyutils
+
+import (
+	"strings"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/logging"
+	"github.com/GoogleCloudPlatform/compute-image-tools/daisy"
+)
+
+// FallbackToPDStandard detects if a workflow fails due to insufficient SSD
+// quota. If so, it re-runs the workflow and modifies the disks to use PD Standard.
+type FallbackToPDStandard struct {
+	logger   logging.Logger
+	fallback bool
+}
+
+// PreRunHook modifies the workflow to use standard disks if `fallback` is true.
+func (t *FallbackToPDStandard) PreRunHook(wf *daisy.Workflow) error {
+	if t.fallback {
+		fallbackToPDStandard(wf)
+	}
+	return nil
+}
+
+// PostRunHook inspects the workflow error to see if it's related to insufficient SSD quota.
+// If so, it requests a retry that will re-run the workflow using standard disks.
+func (t *FallbackToPDStandard) PostRunHook(err error) (wantRetry bool, wrapped error) {
+	if t.fallback {
+		// A fallback has already occurred; don't request retry.
+		t.fallback = false
+	} else if err != nil && strings.Contains(err.Error(), "SSD_TOTAL_GB") {
+		if t.logger != nil {
```

Comment:
Why would logger be `nil`?

Reply:
Removed that check (it simplified testing a bit; but in retrospect it meant we weren't executing the log in tests)